pocs/camera/sdk.py

```diff
@@ -0,0 +1,159 @@
+        # and don't want to do that until after the serial number and port have both been determined
+        # in order to avoid log entries with misleading values. To enable logging during the device
+        # scanning phase use get_root_logger() instead.
+        logger = get_root_logger()
+
+        # The SDK cameras don't generally have a 'port', they are identified by a serial_number,
+        # which is some form of unique ID readable via the camera SDK.
+        kwargs['port'] = None
+        serial_number = kwargs.get('serial_number')
+        if not serial_number:
+            msg = "Must specify serial_number for {}.".format(name)
+            logger.error(msg)
+            raise ValueError(msg)
+
+        if type(self)._driver is None:
+            # Initialise the driver if it hasn't already been done
+            type(self)._driver = driver(library_path=library_path)
+
+        logger.debug("Looking for {} with UID '{}'.".format(name, serial_number))
+
+        if not type(self)._cameras:
+            # No cached camera details, need to probe for connected cameras
+            # This will raise a PanError if there are no cameras.
+            type(self)._cameras = type(self)._driver.get_cameras()
+            logger.debug("Connected {}s: {}".format(name, type(self)._cameras))
+
+        if serial_number in type(self)._cameras:
+            logger.debug("Found {} with UID '{}' at {}.".format(
+                name, serial_number, type(self)._cameras[serial_number]))
+        else:
+            raise error.PanError("Could not find {} with UID '{}'.".format(
+                name, serial_number))
+
+        if serial_number in type(self)._assigned_cameras:
+            raise error.PanError("{} with UID '{}' already in use.".format(
+                name, serial_number))
+
+        type(self)._assigned_cameras.add(serial_number)
+        super().__init__(name, *args, **kwargs)
+        self._address = type(self)._cameras[self.uid]
+        self.connect()
+        assert self.is_connected, error.PanError("Could not connect to {}.".format(self))
+
+        if filter_type:
+            # connect() will have set this based on camera info, but that doesn't know about filters
+            # upstream of the CCD. Can be set manually here, or handled by a filterwheel attribute.
+            self._filter_type = filter_type
+
+        if set_point:
+            self.ccd_set_point = set_point
+            self.ccd_cooling_enabled = True
+
+    def __del__(self):
+        """ Attempt some clean up """
+        with suppress(AttributeError):
+            uid = self.uid
+            type(self)._assigned_cameras.discard(uid)
+            self.logger.debug('Removed {} from assigned cameras list'.format(uid))
+
+    # Properties
+
+    @AbstractCamera.uid.getter
+    def uid(self):
+        """Return unique identifier for camera.
+
+        Need to override this because the base class only returns the 1st
```

Comment:
Still appropriate in the sense that `AbstractCamera` still crops UID to 6 characters because that's what the the Canon gphoto2 `Camera` wants, but all the `Camera` classes that derive from `AbstractSDKCamera` want their full UIDs.

Arguably I should change `AbstractCamera` to uncropped UIDs and override in `AbstractGPhotoCamera` instead though.

Reply:
Doing the latter.

Reply:
Actually, the 6 character limit I think just comes from me and wanting something reasonable to display. The actual id off the Canon's is a full sha256 (or something).

We could consider just always storing the full ID and displaying it shorter where needed.